docs/man/man1/mitsub_vfd.1
.TH mitsub_vfd "1" "2018-10-25" "LinuxCNC Documentation" "HAL Component"

.SH NAME
mitsub_vfd \- HAL non-realtime component for Mitsubishi A500 F500
E500 A500 D700 E700 F700-series VFDs (others may work)
This uses the COMPUTER LINK protocol _not_ MODBUS.
The connection is made through the PU connector.

.SH SYNOPSIS

.B loadrt mitsub_vfd [--baud 4800] [--port /dev/ttyUSB0] name1=number1[,name2=numer2...]
.br
\fBname1\fR is user selectable (usually a description of the controlled device).
.br
\fBnumber1\fR is the slave number that was set on the VFD. Must be two digits (Parameter 117).
.br
\fBname=number\fR can be repeated for multiple VFD's connected together.
.br
\fB--baud\fR is optional as it defaults to 9600, all networked vfds must be set to the same baudrate.
.br
\fB--port\fR is optional as it defaults to ttyS0.

.SH DESCRIPTION

The mitsub_vfd component interfaces a Mitsubishi VFD to LinuxCNC.
The VFD is connected via RS-485 serial to the computer's USB or serial port using a RS-232/RS-485 converter.

.SH HARDWARE SETUP
reference manual 'communication option reference manual' and A500 technical manual for 500 series.
Fr-A700 F700 E700 D700 technical manual for the 700 series.
.br
The inverter must be set manually for communication (you may have to set PR 77 to 1 to unlock PR modification).
.br
You must power cycle the inverter for some of these, e.g. 79.

.TP
\fBVFD INTERNAL PARAMETERS: \fR

\fBPARAMETER 79\fR - 1 or 0

\fBPARAMETER 117\fR station number - 1
.br
(can be optionally set 0 - 31) if component is also set

\fBPARAMETER 118\fR communication speed 96
.br
(can be optionally set 48,96,192 if component is also set)

\fBPARAMETER 119\fR stop bit/data length - 1   
.br      
(8 bits, two stop - don't change)

\fBPARAMETER 120\fR parity - 0
.br
(no parity - don't change)

\fBPARAMETER 121\fR COM tries - 10
.br
(if maximum 10 COM errors then inverter faults- can change.)

\fBPARAMETER 122\fR COM check time interval 9999
.br
(never check - if communication is lost inverter will not know (can change))

\fBPARAMETER 123\fR wait time - 9999
.br
No wait time is added to the serial data frame (don't change).

\fBPARAMETER 124\fR CR selection - 0
.br
don't change

\fBPARAMETER 549\fR communication protocol - 0
.br
computer link protocol - don't change - (not all VFDs have this)

.SH

This driver assumes certain other VFD settings:
-That the motor frequency status is set to show Hertz. 
.br
-That the status bit 3 is up to speed
.br
-That the status bit 7 is alarm

some models (eg E500) cannot monitor status.
.br
You must set set the monitor pin to false.
.br
In this case pins such as up-to-speed,
amps, alarm and status bits are not useful.

.SH PINS
\fB[VFD NAME].fwd\fR (bit, in)::
      forward/reverse pin
.br
\fB[VFD NAME].run\fR (bit, in)::
      run/stop pin
.br
\fB[VFD NAME].debug\fR (bit, in)::
      set debug mode pin
.br
      This will print many messages to the terminal
.br
\fB[VFD NAME].monitor\fR (bit, in)::
      set monitor mode pin
      If false request-status command will not be sent to VFD. Status, amps, power, motor-feedback, and alarm would then not be useful.
.br
\fB[VFD NAME].estop\fR (bit, in)::
      set E-stop mode pin
.br
      This will stop the VFD. Restarting requires the run pin to cycle.

\fB[VFD NAME].fwd\fR (bit, out)::
      up-to-speed status pin
      Motor is at requested speed within VFD's settings tolerance.
.br
\fB[VFD NAME].alarm\fR (bit, out)::
      alarm status pin
.br

\fB[VFD NAME].motor-cmd\fR (float, in)::
      The requested motor speed, is Hertz
.br

\fB[VFD NAME].motor-fb\fR (float, out)::
      The motor feedback speed (from VFD) in Hertz (Hz)
.br
\fB[VFD NAME].motor-amps\fR (float, out)::
      The motor current, in amps
.br
\fB[VFD NAME].motor-power\fR (float, out)::
      The motor power
.br

\fB[VFD NAME].scale-cmd\fR (float, in)::
      Motor command's scale setting defaults to 1
.br
\fB[VFD NAME].scale-cmd\fR (float, in)::
      Motor command's scale setting defaults to 1
.br
\fB[VFD NAME].scale-cmd\fR (float, in)::
      Motor command's scale setting defaults to 1
.br

\fB[VFD NAME].stat-bit-0\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-1\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-2\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-3\fR (bit, out)::
      raw status bit
      set the VFD so this is motor-at-speed status
.br
\fB[VFD NAME].stat-bit-4\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-5\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-6\fR (bit, out)::
      raw status bit
.br
\fB[VFD NAME].stat-bit-7\fR (bit, out)::
      raw status bit
      Set the VFD so this in the alarm bit

.SH SAMPLE HAL
loadusr -Wn coolant mitsub_vfd --port /dev/ttyUSB0 spindle=02 coolant=01
.br
# **************** Spindle VFD setup slave 2 ****************
.br
net spindle-vel-cmd               spindle.motor-cmd
.br
net spindle-cw                    spindle.fwd              
.br
net spindle-on                    spindle.run              
.br
net spindle-at-speed              spindle.up-to-speed
.br
net estop-out                     spindle.estop
.br
#       cmd scaled to RPM (belt/gearbox driven)
.br
setp spindle.scale-cmd .135
.br
#       feedback is in rpm (recipicale of command)
.br
setp spindle.scale-fb 7.411
.br
#       turn on monitoring so feedback works
.br
setp spindle.monitor 1
.br
net spindle-speed-indicator spindle.motor-fb
.br
# *************** Coolant VFD setup slave 1 *****************
.br
net coolant-flood                         coolant.run
.br
net coolant-is-on              coolant.up-to-speed
.br
#       cmd and feedback scaled to hertz
.br
setp coolant.scale-cmd 1
.br
setp coolant.scale-fb 1
.br
#       command full speed
.br
setp coolant.motor-cmd 60
.br
#       allows us to see status
.br
setp coolant.monitor 1
.br
net estop-out                     coolant.estop

.SH ISSUES

Some models, e.g. E500, cannot monitor status, so set the monitor pin to false.
In this case, pins such as up-to-speed, amps, alarm and status bits are not useful.



docs/NEWS
check debian/changelog for recent news

# some of EMC2 History & News
#############################

2009.06.13 - mshaver added emc2/configs/smithy directory and related files
             for Smithy Company's line of cnc machines.

2006.11.17 - removed HAL types s8, u8, s16, and u16.  HAL now uses float
	     for analog values, bit for booleans, and s32 or u32 for
	     integer values.  The only users who will be affected are
	     those with Universal Stepper Controller cards.  See item 9 at
	     http://wiki.linuxcnc.org/cgi-bin/emcinfo.pl?UpdatingConfigurationsForDevelopmentVersions

2006.11.15 - merged classicladder 0.7.100

2006.11.04 - added a feed-hold HAL pin.  Added M codes to enable/disable
	     feed override, spindle override, feed hold, and adaptive feed
	     individually.  Feed override is M50, spindle override is M51,
	     adaptive feed is M52, and feed hold is M53.  G50/51 are no
	     longer used to enable/disable adaptive feed, use M52 instead.

2006.10.24 - added support for block delete as defined by Section 2.2.2
             in the RS274NGC_3 specs.

2006.10.24 - added support for optional program stop as defined by Section 
             2.2.3 in the RS274NGC_3 specs.

2006.10.14 - manpages have been created for most rtapi and hal calls

2006.10.01 - added a userspace simulator which includes rtapi threads.
	     Use "--enable-simulator" to enable this mode, which cannot
	     be used to control real hardware

2006.09.27 - added support to optionally save joint positions between
             runs ([TRAJ] POSITION_FILE = position.txt)

2006.09.26 - added support to specify UNITS in literal in the ini. 
             (e.g. [TRAJ] LINEAR_UNITS = mm )

2006.09.09 - fixed bug #1384883, optional stop was not optional.

2006.08.09 - AXIS is now a part of EMC2, not a separate source download.

2006.08.07 - Added a spindle-speed override. This allows to override the spindle
	     speed during a program run. The MIN_SPINDLE_OVERRIDE and 
	     MAX_SPINDLE_OVERRIDE in the ini assure that speed doesn't get 
	     reduced or increased too much. Both values are in %.

2006.08.06 - Documentation can be built from source with the 'make docs' target

2006.06.11 - added realtime support for jogwheels.  Three HAL pins per axis:
	     one accepts raw counts from the wheel, one sets the distance moved
	     per count, and one enables that axis to move.

2006.05.17 - added 'adaptive feed', a HAL pin that can be used to scale the
	     feedrate in real time.  Simular to feedrate override, but can
	     only vary from 0.0 to 1.0, and reacts much faster.  Intended 
	     for use with EDM and other applications that require closed 
	     loop control of feedrate.

2006.03.18 - A new pin "count" exists for freqgen.  It is the same as the
             internal symbol "count" shown in the hal introduction.pdf,
             or the result of up OR down if there was an "OR" component.

2006.03.16 - Modified the motion control module to behave more like other
             HAL modules - it no longer adds its functions to threads
             automatically.  All the sample configs have been changed to
             explicitly add the functions, users who have custom configs
             will need to do the same.

2006.03.12 - If a copy of axis 1.3a0 or newer is detected inside src/
             it will be automatically built and installed for
             --enable-run-in-place use

2006.03.10 - minor bugfix in tkemc for jogging

2006.03.10 - added genhexkins, with a minitetra sample config.
             This proves EMC2 works just as good with hexapods as EMC1 did.

2006.03.10 - added VTI support & sample configs

2006.03.07 - lots of fixes to the TP (most were wrong in the old one as well)

2006.03.05 - new Trajectory Planner

2006.02.20 - added a manpage for the runscript.

2006.02.12 - Fixed a bug that was causing the RT code to run very slowly,
             forcing users to set BASE_PERIOD very long.

2006.02.10 - The build process has been replaced, no more recursive $(MAKE)

2006.01.28 - fixed bug #1205237, mode switching sometimes hung

2006.01.08 - M62/M63 digital outputs synchronous to motion
             M64/M65 digital outputs that get set right away

2006.01.06 - work on halconfig.tcl, it is used to display/configure HAL data
             requires the BWidget package, realtime, and HAL to run.

2006.01.04 - fix bug #1387740, a motion that goes nowhere clears feedrate

2006.01.01 - the config picker has i18n support, a German translation exists


src/hal/user_comps/mitsub_vfd.py
#!/usr/bin/env python3
#    This is a component of linuxcnc
#    mitsub_vfd Copyright 2017 Chris Morley
#
#
#    This program is free software; you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation; either version 2 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program; if not, write to the Free Software
#    Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.

# user space component for controlling a misubishi inverter over the serial port using rs485 standard
# specifically the A500 F500 E500 A500 D700 E700 F700 series - others may work or need adjustment
# tested on A500 E500 and E700
# I referenced manual 'communication option reference manual' and A500 technical manual for 500 series.
# 'Fr-A700 F700 E700 D700 technical manual' for the 700 series
#
# The inverter must be set manually for communication ( you may have to set PR 77 to 1 to unlock PR modification )
# must power cycle the inverter for some of these to register eg 79
# PR 79 - 1 or 0
# PR 117 station number - 1               (can be optionally set 0 - 31) if component is also set
# PR 118 communication speed 96           (can be optionally set 48,96,192) if component is also set
# PR 119 stop bit/data length - 1         8 bits, two stop (don't change)
# PR 120 parity - 0                       no parity (don't change)
# PR 121 COM tries - 10                   if 10 (maximuim) COM errors then inverter faults (can change)
# PR 122 COM check time interval 9999     (never check) if communication is lost inverter will not know (can change)
# PR 123 wait time - 9999 -               no wait time is added to the serial data frame (don't change)
# PR 124 CR selection - 0                 don't change
# PR 549 communication protocol - 0        not all VFDs has this

import time,hal
import serial
import traceback
class mitsubishi_serial:

    def __init__(self,vfd_names=[['mitsub_vfd','00']],baudrate=9600,port='/dev/ttyUSB0'):
        try:
            self.ser = serial.Serial(
            port,
            baudrate,
            parity=serial.PARITY_NONE,
            stopbits=serial.STOPBITS_TWO,
            bytesize=serial.EIGHTBITS
            )
            self.ser.open()
            self.ser.isOpen()
        except Exception as e:
            try:
                self.ser.close()
                self.ser.open()
            except:
                print("ERROR : mitsub_vfd - No serial interface found at %s\nError: %s"% (port,e))
                pass
            #raise SystemExit
        print("Mitsubishi VFD serial computer link has loaded")
        print("Port: %s,\nbaudrate: %d\n8 data bits, no parity, 2 stop bits\n"%(port,baudrate))

        self.h=[]
        self.comp_names = vfd_names
        for index,name in enumerate(self.comp_names):
            #print index,' NAME:',name[0],' SLAVE:',name[1]
            c = hal.component(name[0])
            c.newpin("fwd", hal.HAL_BIT, hal.HAL_IN)
            c.newpin("run", hal.HAL_BIT, hal.HAL_IN)
            c.newpin("up-to-speed", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("alarm", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("debug", hal.HAL_BIT, hal.HAL_IN)
            c.newpin("monitor", hal.HAL_BIT, hal.HAL_IN)
            c.newpin("motor-cmd", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("motor-fb", hal.HAL_FLOAT, hal.HAL_OUT)
            c.newpin("motor-amps", hal.HAL_FLOAT, hal.HAL_OUT)
            c.newpin("motor-volts", hal.HAL_FLOAT, hal.HAL_OUT)
            c.newpin("motor-power", hal.HAL_FLOAT, hal.HAL_OUT)
            c.newpin("motor-user", hal.HAL_FLOAT, hal.HAL_OUT)
            c.newpin("scale-cmd", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("scale-fb", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("scale-amps", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("scale-volts", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("scale-power", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("scale-user", hal.HAL_FLOAT, hal.HAL_IN)
            c.newpin("estop", hal.HAL_BIT, hal.HAL_IN)
            c.newpin("stat-bit-0", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-1", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-2", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-3", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-4", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-5", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-6", hal.HAL_BIT, hal.HAL_OUT)
            c.newpin("stat-bit-7", hal.HAL_BIT, hal.HAL_OUT)
            # set reasonable defaults
            c['scale-cmd'] = 1
            c['scale-fb'] = 1
            c['scale-amps'] = 1
            c['scale-volts'] = 1
            c['scale-power'] = 1
            c['scale-user'] = 1
            c['fwd'] = 1
            # flags for each device
            self['last_run%d'%index] = c['run']
            self['last_fwd%d'%index] = c['fwd']
            self['last_cmd%d'%index] = c['motor-cmd']
            self['last_monitor%d'%index] = c['monitor']
            self['last_estop%d'%index] = c['estop']
            #add device to component reference variable
            self.h.append(c)
            print("Mitsubishi %s VFD: slave# %s added\n"%(name[0],name[1]))

        # only issue ready when all the components are ready
        for i in self.h:
            i.ready()
        self.set_special_monitor()

    def loop(self):
        cmd = data = out = temp=''
        while 1:
            try:
              for index,ids in enumerate(self.comp_names):
                self.slave_num = ids[1]
                # MONITOR for up-to-speed, alarms and running frequency
                # 7A is the address for 8 status bits ( b0 - b8 )
                # These bits are configurable from the panel. We assume bit 3 is up to speed
                # and bit 7 is alarm
                # the returned data is 2 characters of hex
                # we convert that to binary and the mask for the bits we want
                if self.h[index]['monitor']:
                    while self.ser.inWaiting() > 0:
                        raw = self.ser.read(1)
                    word = self.prepare_data("7A",None)
                    out = temp = ''
                    self.ser.write(word)
                    time.sleep(.05)
                    string,chr_list,chr_hex = self.poll_output()
                    #print 'DEBUG: ',chr_list,chr_hex
                    if chr_list != '':
                        #print string
                        try:
                            binary = "{0:#010b}".format(int(string[3:5],16))
                        except:
                            binary = '0b000000'
                        self.h[index]['stat-bit-0'] = int(binary,2) & 1
                        self.h[index]['stat-bit-1'] = int(binary,2) & 2
                        self.h[index]['stat-bit-2'] = int(binary,2) & 4
                        self.h[index]['stat-bit-3'] = self.h[index]['up-to-speed'] = int(binary,2) & 8
                        self.h[index]['stat-bit-4'] = int(binary,2) & 16
                        self.h[index]['stat-bit-5'] = int(binary,2) & 32
                        self.h[index]['stat-bit-6'] = int(binary,2) & 64
                        self.h[index]['stat-bit-7'] = self.h[index]['alarm'] = int(binary,2) & 128
                        if self.h[index]['debug'] and 1==2:
                            print('monitor operation:',binary,temp,temp[3:5],len(temp))

                    # 6F is the address for running motor frequency status
                    # it returns 4 characters of hex
                    # we convert to decimal and multiply by .01 for hertz and by user scale-fb
                    # for arbrtrary units. This does require scale to be set to something besides 0!
                    # we assume the inverter is set to show running hertz (it's configurable in the VFD)
                    word = self.prepare_data("6F",None)
                    out = temp = ''
                    self.ser.write(word)
                    time.sleep(.05)
                    string,chr_list,chr_hex = self.poll_output()
                    if self.h[index]['debug']:
                        print('DEBUG: ',chr_list,chr_hex)
                    if chr_list != '':
                        decimal = int(string[3:7],16)
                        self.h[index]["motor-fb"] = decimal *.01 * self.h[index]["scale-fb"]
                        if self.h[index]['debug'] and 1==2:
                            print('monitor frequency:',decimal,string,string[3:7], len(string))

                # amps
                    word = self.prepare_data("70",None)
                    out = temp = ''
                    self.ser.write(word)
                    time.sleep(.05)
                    string,chr_list,chr_hex = self.poll_output()
                    if self.h[index]['debug']:
                        print('DEBUG: ',chr_list,chr_hex)
                    if chr_list != '':
                        decimal = int(string[3:7],16)
                        self.h[index]["motor-amps"] = decimal *.01 * self.h[index]["scale-amps"]
                        if self.h[index]['debug'] and 1==2:
                            print('monitor amps:',decimal,string,string[3:7], len(string))

                # volts
                    word = self.prepare_data("71",None)
                    out = temp = ''
                    self.ser.write(word)
                    time.sleep(.05)
                    string,chr_list,chr_hex = self.poll_output()
                    if self.h[index]['debug']:
                        print('DEBUG: ',chr_list,chr_hex)
                    if chr_list != '':
                        decimal = int(string[3:7],16)
                        self.h[index]["motor-volts"] = decimal *.01 * self.h[index]["scale-volts"]
                        if self.h[index]['debug'] and 1==2:
                            print('monitor volts:',decimal,string,string[3:7], len(string))
                # Power
                    self.h[index]["motor-power"] = self.h[index]["motor-volts"] * self.h[index]["motor-amps"] * 2.7

                # special user selected monitor
                    word = self.prepare_data("72",None)
                    out = temp = ''
                    self.ser.write(word)
                    time.sleep(.05)
                    string,chr_list,chr_hex = self.poll_output()
                    if self.h[index]['debug']:
                        print('DEBUG: ',chr_list,chr_hex)
                    if chr_list != '':
                        decimal = int(string[3:7],16)
                        self.h[index]["motor-user"] = decimal * self.h[index]["scale-user"]
                        if self.h[index]['debug'] and 1==2:
                            print('monitor user selectable:',decimal,string,string[3:7], len(string))

                # STOP ON ESTOP
                # if ESTOP is false it stops the output
                # when ESTOP is reset the run command must be re-issued (cycled false to true) to start motor 
                if not self['last_estop%d'%index] == self.h[index]['estop']:
                    if not self.h[index]["estop"]:
                        cmd = "FA";data ="00"
                        word = self.prepare_data(cmd,data)
                        self.ser.write(word)
                        time.sleep(.05)
                        self['last_estop%d'%index] = self.h[index]['estop']
                        print("**** Mitsubishi VFD: %s stopped due to Estop Signal"% ids[0])
                        continue
                    else:
                        # reset VFD after estop
                        cmd = "FD";data = None
                        word = self.prepare_data(cmd,data)
                        self.ser.write(word)
                        time.sleep(.05)
                    
                    print("**** Mitsubishi VFD: Estop cleared - Must re-issue run command to start %s." % ids[0])
                    self['last_estop%d'%index] = self.h[index]['estop']

                # SET RUN AND DIRECTION
                # address FA sets the start and direction
                # it expects a 2 character hex representing a 8 bit (b0 - b7) binary number
                # bit 1 sets forward, 4 sets reverse, 0 stop
                # depending on the inverter and options other bits are possible,
                # but these three are consistent
                if not self['last_run%d'%index] == self.h[index]['run'] or not self['last_fwd%d'%index] == self.h[index]['fwd']:
                    if self.h[index]['run']:
                        if self.h[index]['fwd']:
                            cmd = "FA";data ="02"
                        else:
                            cmd = "FA";data ="04"
                    else:
                        cmd = "FA";data ="00"
                    word = self.prepare_data(cmd,data)
                    self.ser.write(word)
                    time.sleep(.05)
                    self['last_run%d'%index] = self.h[index]['run']
                    self['last_fwd%d'%index] = self.h[index]['fwd']
                    if self.h[index]['debug']:
                        string,chr_list,chr_hex = self.poll_output()
                        print('DEBUG: ',chr_list,chr_hex)

                # SET cmd
                # address ED is for setting the running frequency
                # it expects 4 characters of hex representing frequency in .01 hertz units
                # we internally scale it by 100 to make it 1 hertz units and by user scale
                # for arbrtrary units. This does require scale to be set to something besides 0!
                if not self['last_cmd%d'%index] == self.h[index]['motor-cmd']:
                    freq = int(abs(self.h[index]['motor-cmd']*100*self.h[index]['scale-cmd']))
                    if freq > 40000: freq = 40000
                    if freq < 0: freq = 0
                    self['last_cmd%d'%index] = self.h[index]['motor-cmd']
                    # send frequency command
                    cmd="ED";data="%0.4X"%freq
                    word = self.prepare_data(cmd,data)
                    self.ser.write(word)
                    time.sleep(.05)
                    if self.h[index]['debug']:
                        string,chr_list,chr_hex = self.poll_output()
                        print('DEBUG: ',chr_list,chr_hex)

            except KeyboardInterrupt:
                    self.kill_output()
                    raise
            except Exception as e:
                    print("error",ids)
                    print(sys.exc_info()[0])
                    print (e)

    # defaults to power in kw
    def set_special_monitor(self, option = '0E'):
        cmd="F3"
        for index,ids in enumerate(self.comp_names):
            self.slave_num = ids[1]
            word = self.prepare_data(cmd,option)
            self.ser.write(word)
            time.sleep(.05)

    def kill_output(self):
        cmd = "FA";data ="00"
        for index,ids in enumerate(self.comp_names):
            self.slave_num = ids[1]
            word = self.prepare_data(cmd,data)
            self.ser.write(word)
            time.sleep(.05)
            print('Mitsub VFD: Kill-> ', ids[0])

    def prepare_data(self,command ='E1',data= '07AD'):
        combined = self.slave_num+command  +'1'
        if not data == None:
            combined += data
        s=0
        for i in range(0,len(combined)):
            letter = combined[i]
            s=s+ord(letter.upper())
        converted_data = chr(0x5) + combined + hex(s)[-2:-1].upper() + hex(s)[-1:].upper()
        return bytes(converted_data, 'utf-8')

    def poll_output(self):
        string = chr_list_out = hex_out = ''
        while self.ser.inWaiting() > 0:
            raw = self.ser.read(1).decode()
            chr_list_out += raw+','
            string += raw
            hex_out += hex(ord(raw))
            hex_out +=' '
        if hex_out != '':
            answ=''
            if chr_list_out[0] == chr(0x6):
                answ = 'ackg'
            if chr_list_out[0] == chr(0x15):
                answ = 'error'
            if chr_list_out[0] == chr(0x2):
                answ = 'checksome error'
            #print 'slave:',chr_list_out[2]+chr_list_out[4],answ
            return string,chr_list_out,hex_out
        return '','',''

    def __getitem__(self, item):
        return getattr(self, item)
    def __setitem__(self, item, value):
        return setattr(self, item, value)

if __name__ == "__main__":
    import getopt,sys
    letters = 'p:b:h' # the : means an argument needs to be passed after the letter
    keywords = ['port=', 'baud=' ] # the = means that a value is expected after 
    # the keyword
    opts, extraparam = getopt.getopt(sys.argv[1:],letters,keywords) 
    # starts at the second element of argv since the first one is the script name
    # extraparms are extra arguments passed after all option/keywords are assigned
    # opts is a list containing the pair "option"/"value"
    port='/dev/ttyS0'
    device_ids=[]
    baud=9600

    for o,p in opts:
      if o in ['-p','--port']:
         port = p
      elif o in ['-b','--baud']:
         baud = p
      elif o in ['-h','--help']:
        print('Mitsubishi VFD COMPUTER-LINK interface')
        print('This does NOT use the MODBUS protocol.')
        print(' User space component for controlling a mitsubishi inverter over the serial port using the rs485 standard')
        print(' specifically the A500 F500 E500 A500 D700 E700 F700 series - others may work or need small adjustments')
        print(''' I referenced manual 'communication option reference manual' and A500 technical manual for 500 series.''')
        print(''' 'Fr-A700 F700 E700 D700 technical manual' for the 700 series''')
        print() 
        print(' The inverter must be set manually for communication ( you may have to set PR 77 to 1 to unlock PR modification )')
        print(' You must power cycle the inverter for some of these to register eg 79')
        print(' PR 79 - 1 or 0                          sets the inverter to respond to the PU/computer-link')
        print(' PR 117 station number (slave) - 1       can be optionally set 0 - 31 if component is also set')
        print(' PR 118 communication speed 96           baud rate, can be optionally set 48,96,192 if component is also set')
        print(''' PR 119 stop bit/data length - 1         8 bits, two stop (don't change)''')
        print(''' PR 120 parity - 0                       no parity (don't change)''')
        print(' PR 121 COM tries - 10                   if 10 (maximuim) COM errors then inverter faults (can change)')
        print(' PR 122 COM check time interval 9999     (never check) if communication is lost inverter will not know (can change)')
        print(''' PR 123 wait time - 9999 -               no wait time is added to the serial data frame (don't change)''')
        print(''' PR 124 CR selection - 0                 don't change''')
        print(''' PR 549 communication protocol - 0        not all VFDs has this''')
        print('''
This driver assumes certain other VFD settings:
-That the  motor frequency status is set to show herts.
-That the status bit 3 is up to speed
-That the status bit 7 is alarm
''')
        print()
        print('''some models (eg E500) cannot monitor status -set the monitor pin to false
in this case pins such as up-to-speed, amps, alarm and status bits are not useful.
''')
        print('''HAL command used to load: ''')
        print('''loadusr mitsub_vfd --baud 4800 --port /dev/ttyUSB0 NAME=SLAVE_NUMBER 
        -NAME is user selectable (usually a description of controlled device)
        -SLAVE_NUMBER is the slave number that was set on the VFD
        -NAME=SLAVE_NUMBER can be repeated for multiple VFD's connected together
        --baud is optional as it defaults to 9600
        all networked vfds must be set to the same baudrate
        --port is optional as it defaults to ttyS0''')
        print()
        print(''' Sample linuxcnc code
loadusr -Wn coolant mitsub_vfd spindle=02 coolant=01
# **************** Spindle VFD setup slave 2 *********************
net spindle-vel-cmd               spindle.motor-cmd
net spindle-cw                    spindle.fwd              
net spindle-on                    spindle.run              
net spindle-at-speed              spindle.up-to-speed
net estop-out                     spindle.estop
#       cmd scaled to RPM
setp spindle.scale-cmd .135
#       feedback is in rpm
setp spindle.scale-fb 7.411
setp spindle.monitor 1

net spindle-speed-indicator spindle.motor-fb        gladevcp.spindle-speed
# *************** Coolant vfd setup slave 3 ***********************
net coolant-flood                         coolant.run
net coolant-is-on              coolant.up-to-speed   gladevcp.coolant-on-led
#       cmd and feedback scaled to hertz
setp coolant.scale-cmd 1
setp coolant.scale-fb 1
#       command full speed
setp coolant.motor-cmd 60
#       allows us to see status
setp coolant.monitor 1
net estop-out                     coolant.estop
''')
        sys.exit(0)
    if extraparam:
        for dids in extraparam:
            device_ids.append(dids.split('='))
    else:
        device_ids=[['mitsub_vfd','00']]
    print(port,baud)
    # Info gathered now use them
    try:
        app = mitsubishi_serial(vfd_names=device_ids,baudrate=int(baud),port=port)
        app.loop()
    except KeyboardInterrupt:
        sys.exit(0)
    else:
            exc_type, exc_value, exc_traceback = sys.exc_info()
            formatted_lines = traceback.format_exc().splitlines()
            print()
            print("**** Mitsub_vfd debugging:",formatted_lines[0])
            traceback.print_tb(exc_traceback, limit=1, file=sys.stdout)
            print(formatted_lines[-1])
